common/util.go

```diff
@@ -521,8 +521,12 @@ func CreateHistoryStartWorkflowRequest(
 	histRequest.FirstDecisionTaskBackoffSeconds = Int32Ptr(firstDecisionTaskBackoffSeconds)
 
 	if startRequest.RetryPolicy != nil && startRequest.RetryPolicy.GetExpirationIntervalInSeconds() > 0 {
-		expirationInSeconds := startRequest.RetryPolicy.GetExpirationIntervalInSeconds() + firstDecisionTaskBackoffSeconds
+		expirationInSeconds := Int32Default(startRequest.ExecutionStartToCloseTimeoutSeconds)
+		if startRequest.RetryPolicy.GetExpirationIntervalInSeconds() > expirationInSeconds {
+			expirationInSeconds = startRequest.RetryPolicy.GetExpirationIntervalInSeconds()
+		}
```

Comment:
- If execution timeout is 10h, retry expiration interval is 1h and workflow fails after 2h. With the new logic, I think the workflow will be retried as the expiration is 10h? But looks like we should not do the retry since the retry expiration has passed?

- For later attempts of the workflow, I think we will still encounter the situation where workflow timeout time is after retry expiration time and workflow will timeout earlier by the retry expiration.

- The second point means we kinda have an inconsistency of behavior for first attempts and later attempts since we guarantee the first attempt won't be cut off by retry expiration but later attempts will. I guess it would be better to keep the behavior consistent?

Reply:
@yycptt 

`If execution timeout is 10h, retry expiration interval is 1h and workflow fails after 2h. With the new logic, I think the workflow will be retried as the expiration is 10h? But looks like we should not do the retry since the retry expiration has passed?`
Hm, maybe I misunderstood what the RetryPolicy.ExpirationIntervalInSeconds means; My interpretation is that if the workflow retries, but in the retry takes longer than 2h, it will timeout. Are you saying that the retry expiration means in the case when 2h is pass we will not attempt a retry? 

`For later attempts of the workflow, I think we will still encounter the situation where workflow timeout time is after retry expiration time and workflow will timeout earlier by the retry expiration.`
I think related to my first point -- if the workflow is retried, should we not timeout earlier by retry expiration? And a followup question -- should the the RetryPolicy's expiration interval be absolute or per retry (i.e. if its 1h does that mean a) each retry should timeout after 1h or b) if 1h is past from StartWorkflow no retries will be attempted)

As a followup question for subsequent retries, I'm seeing that the method I modified `CreateHistoryStartWorkflowRequest` is called in `startWorkflowWithRetry`, `StartWorkflowExecution`, `SignalWithStartWorkflowExecution` and `terminateAndStartWorkflow`, If i were to make it consistent such that retries will not time out earlier -- do you know which other method should I make the modification in?

Reply:
- yeah, that's my understanding of the ticket.  
- For the second and third comments, my main point is it looks like the behavior will be inconsistent between the first and later attempts. I am not leaning towards any of them though, both sounds reasonable to me. 
- This is just my thought. We can discuss with the team to reach a conclusion.

My understanding it the expiration interval is absolute and is calculated and converted to expiration timestamp once when starting the workflow (or the beginning of each run for cron). Then the expiration timestamp will be used across all retires.

You may want to look into how workflow timeout task is generated: https://github.com/uber/cadence/blob/master/service/history/execution/mutable_state_task_generator.go#L138
and how the expiration timestamp is carried over to the next run:
https://github.com/uber/cadence/blob/master/service/history/execution/mutable_state_builder.go#L1700